File: skills/project-context/hooks/pre-compact-backup.py
#!/usr/bin/env python3
"""PreCompact safety net.

Before context compaction, copy the project's project-context.json to a
timestamped backup so a mechanical snapshot survives even if the model did not
refresh it in time. Does NOT block compaction — the meaningful refresh is driven
preemptively by ctx-guard (via the statusline sensor); this is the last-resort
snapshot. Silent no-op when the file is absent. Keeps the most recent 10 backups.
"""
import json
import os
import shutil
import sys
import time


def main():
    try:
        data = json.load(sys.stdin)
    except Exception:
        data = {}

    cwd = data.get("cwd") or os.getcwd()
    src = os.path.join(cwd, "meta", "structures", "project-context.json")
    if not os.path.isfile(src):
        return

    backup_dir = os.path.join(cwd, "meta", "structures", ".pre-compact-backups")
    try:
        os.makedirs(backup_dir, exist_ok=True)
        dst = os.path.join(backup_dir, "project-context-" + time.strftime("%Y%m%d-%H%M%S") + ".json")
        shutil.copy2(src, dst)
    except Exception:
        return  # never break compaction

    # Retain only the 10 most recent backups.
    try:
        backups = sorted(f for f in os.listdir(backup_dir) if f.endswith(".json"))
        for old in backups[:-10]:
            os.remove(os.path.join(backup_dir, old))
    except Exception:
        pass


if __name__ == "__main__":
    main()

File: skills/project-structure/hooks/pre-compact-backup.mjs
#!/usr/bin/env node
// PreCompact safety net for project-structure.
//
// Before context compaction, copy project-structure.json to a timestamped
// backup so a mechanical snapshot survives even if the model didn't refresh the
// map in time. Does NOT block compaction. Silent no-op when absent. Keeps the
// 10 most recent backups. (The history .jsonl is append-only, so it doesn't need
// backing up — it can't lose data.)
import fs from "node:fs";
import path from "node:path";

function stamp() {
  const d = new Date();
  const p = (n) => String(n).padStart(2, "0");
  return (
    `${d.getFullYear()}${p(d.getMonth() + 1)}${p(d.getDate())}-` +
    `${p(d.getHours())}${p(d.getMinutes())}${p(d.getSeconds())}`
  );
}

let input = "";
process.stdin.on("data", (c) => (input += c));
process.stdin.on("end", () => {
  let data = {};
  try {
    data = JSON.parse(input);
  } catch {}

  const cwd = data.cwd || process.cwd();
  const src = path.join(cwd, "meta", "structures", "project-structure.json");
  if (!fs.existsSync(src)) return;

  const dir = path.join(cwd, "meta", "structures", ".pre-compact-backups");
  try {
    fs.mkdirSync(dir, { recursive: true });
    fs.copyFileSync(src, path.join(dir, `project-structure-${stamp()}.json`));
  } catch {
    return; // never break compaction
  }

  // Retain only the 10 most recent structure backups.
  try {
    const backups = fs
      .readdirSync(dir)
      .filter((f) => f.startsWith("project-structure-") && f.endsWith(".json"))
      .sort();
    for (const old of backups.slice(0, -10)) {
      fs.unlinkSync(path.join(dir, old));
    }
  } catch {}
});
